### PR TITLE
DIS: store degree progress in backend

### PR DESCRIPTION
**Pull request by mpark63**
todos before merge: 
- [x] update from dev 
- [ ] jest tests
- [x] update majors.js
- [x] clean up data folder
- [ ] fix bugs 
- [x] add comments
- [ ] test deployment 
  - [x] add / delete / move course 
  - [x] distribution credit update 
  - [x] fine requirement credit update 
  - [x] add / remove plan 
  - [x] add / remove major(s)
  - [ ] plan sharing consistency
  - [x] reviewer view  

todos after merge: 
- [ ] add majors 
- [ ] schema changes (`updateModels()` in `migrateDistributions.js`)
  - [ ] rename `Course.area` to areas 
  - [ ] add distribution_ids and finereq_ids to `Course` 
  - [ ] rename `Plan.majors` to `Plan.major_ids`
  - [ ] remove `Plan.distribution_ids`
  - [ ] `Year.plan_id` should be ObjectId and not array 
- [ ] delete any existing dist/finereqs
- [ ] set `Course.department` field if DNE (`setDepartmentInCourses()`)
- [ ] set `Course.tags` field if DNE (`setTagInCourses()`)
- [ ] create new distribution documents to every plan (`addDistributions()`)
- [ ] update every distribution (`addAllCourses()`)

### CODE REVIEW

File: data/cacheSIS.js
Comment (mpark63) [inline data/cacheSIS.js:1]: obsolete

File: data/courseSamples.js
Comment (mpark63) [inline data/courseSamples.js:1]: obsolete

File: data/distributionSamples.js
Comment (mpark63) [inline data/distributionSamples.js:1]: obsolete

File: data/major.js
Comment (mpark63) [inline data/major.js:1]: obsolete

File: data/majors.js
Comment (mpark63) [inline data/majors.js:1]: added all the majors here; deleted all from majors.tsx in frontend 

from now one changes need to be made here and run `updateMajors.js`

File: data/migrateDistributions.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        count++;
+      }
+    }
+    console.log(count, ' year documents modified.');
+  } catch (err) {
+    console.log(err);
+  }
+}
+
+async function setDepartmentInCourses() {
```
Comment (mpark63) [inline data/migrateDistributions.js:66]: don't remember why this was needed but seems to have been an issue before so will run for good measure. 

basically sets course department field by looking up course in SISCourseV

File: data/migrateDistributions.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      console.log('matched: %d', res.length);
+      console.log('updated from SISCourseV: %d', updated);
+      console.log('updated unspecified: %d', res.length - updated);
+    })
+    .catch((err) => {
+      console.log(err);
+    });
+}
+
+async function setTagInCourses() {
```
Comment (mpark63) [inline data/migrateDistributions.js:97]: same as above, 

set Tag if not set already

File: data/migrateDistributions.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+      console.log('matched: %d', res.length);
+      console.log('updated from SISCourseV: %d', updated);
+    })
+    .catch((err) => {
+      console.log(err);
+    });
+}
+
+// delete all Distributions Documents
+async function deleteDistributionsDocuments() {
```
Comment (mpark63) [inline data/migrateDistributions.js:125]: delete all distributions document in prep for migration

File: data/migrateDistributions.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      await addPlanDistributions(plan);
+    }
+    console.log(plans.length);
+  });
+}
+
+// add courses to distributions of all plans
+async function addAllCourses() {
+  const plans = await Plans.find({});
+  for (let plan of plans) {
```
Comment (mpark63) [inline data/migrateDistributions.js:162]: add distributions to every plan 

update every distribution with course

File: data/replaceOldMajors.js [hunk 5]
```diff
@@ -2,7 +2,8 @@ import { connect } from './db.js';
 import User from '../model/User.js';
 import Plan from '../model/Plan.js';
 
-async function replaceOldWithNew() {
+// replace old major names with new ones
+async function renameOldWithNew() {
```
Comment (mpark63) [inline data/replaceOldMajors.js:6]: renamed for clarity

File: data/updateMajors.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+import * as db from './db.js';
+import { allMajors } from './majors.js';
+import Majors from '../model/Major.js';
+
+// addAllMajors();
+updateAllMajors();
+
+// add majors in allMajors array if not already in production DB
+async function addAllMajors() {
```
Comment (mpark63) [inline data/updateMajors.js:15]: add new majors

File: data/updateMajors.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      console.log(`${major._id} already exists!`);
+    }
+  }
+  console.log(
+    `Done adding ${new_count} new majors! Check DB to confirm the documents have been added to the collection`,
+  );
+}
+
+// add new majors and update existing ones in production DB
+async function updateAllMajors() {
```
Comment (mpark63) [inline data/updateMajors.js:37]: add new majors and update existing

Comment (mpark63) [reply data/updateMajors.js:37]: need to run this one to add majors in majors.js

File: model/Course.js [hunk 8]
```diff
@@ -21,7 +21,7 @@ const courseSchema = new Schema({
   number: String,
   department: String,
   tags: [{ type: String }],
-  area: String,
+  areas: { type: String },
```
Comment (mpark63) [inline model/Course.js:24]: felt it is more reasonable to call it areas

File: model/Course.js [hunk 9]
```diff
@@ -30,7 +30,10 @@ const courseSchema = new Schema({
   isTransfer: { type: Boolean, default: false },
   ratings: Array,
   distribution_ids: [
-    { type: Schema.Types.ObjectId, ref: 'Distribution', required: true },
+    { type: Schema.Types.ObjectId, ref: 'Distribution', default: [] },
+  ],
+  fineReq_ids: [
+    { type: Schema.Types.ObjectId, ref: 'FineRequirement', default: [] },
```
Comment (mpark63) [inline model/Course.js]: take note of which dist / finereqs a. course is part of 

for future projects like looking up what distributions a course satisfies or vice versa--all the courses that satisfy a distribution

File: model/Distribution.js [hunk 10]
```diff
@@ -10,14 +10,23 @@ const Schema = mongoose.Schema;
...
+  required_credits: { type: Number, required: true },
+  description: { type: String, required: true },
+  criteria: { type: String, default: '' },
+  min_credits_per_course: { type: Number, required: true },
+  user_id: { type: String, required: true },
+  plan_id: { type: Schema.Types.ObjectId, ref: 'Plan', required: true },
+  major_id: { type: String, required: true },
+  fineReq_ids: [
+    { type: Schema.Types.ObjectId, ref: 'FineRequirement', default: [] },
+  ],
```
Comment (mpark63) [inline model/Distribution.js]: new schema so no need for migration

File: model/Major.js [hunk 11]
```diff
@@ -25,19 +26,17 @@ const majorSchema = new Schema({
...
+      double_count: [{ type: String }], // whitelisted distributions that can double count; if undefined, all whitelisted
       fine_requirements: [
         {
           description: { type: String, required: true },
           required_credits: { type: Number, required: true },
-          criteria: { type: String, required: true },
-          exception: { type: String },
-          exclusive: { type: Boolean, default: false },
+          criteria: { type: String, default: '' },
+          double_count: [{ type: String }],
```
Comment (mpark63) [inline model/Major.js:39]: template from which we make FineRequirement documents; missing some attributes hence not the same

File: model/Plan.js [hunk 12]
```diff
@@ -8,9 +8,8 @@ const Schema = mongoose.Schema;
 const planSchema = new Schema(
   {
     name: { type: String, required: true },
-    majors: { type: [String] },
+    major_ids: [{ type: String, required: true }],
```
Comment (mpark63) [inline model/Plan.js:11]: renamed for consistency with other fields

File: model/Year.js [hunk 13]
```diff
@@ -4,10 +4,10 @@ const Schema = mongoose.Schema;
 const yearSchema = new Schema({
   name: { type: String, required: true }, //freshman
   courses: [{ type: Schema.Types.ObjectId, ref: 'Course' }],
-  plan_id: [{ type: Schema.Types.ObjectId, ref: 'Plan', required: true }],
+  plan_id: { type: Schema.Types.ObjectId, ref: 'Plan', required: true },
```
Comment (mpark63) [inline model/Year.js:7]: should not be an array

File: routes/course.js [hunk 14]
```diff
@@ -133,16 +130,28 @@ router.post('/api/courses', auth, async (req, res) => {
         });
       }
     }
-    // create course and update distributiosn
+    // create course
     const retrievedCourse = await Courses.create(course);
+    // update plan's distribution objs by major
```
Comment (mpark63) [inline routes/course.js:136]: try adding course to every distribution in plan (for one or more majors)

File: routes/course.js [hunk 15]
```diff
@@ -279,14 +288,7 @@ router.delete('/api/courses/:course_id', auth, async (req, res) => {
...
     ).exec();
-    for (let id of course.distribution_ids) {
-      const distribution = await Distributions.findByIdAndUpdate(
-        id,
-        { $pull: { courses: c_id } },
-        { new: true, runValidators: true },
-      ).exec();
-      await distributionCreditUpdate(distribution, course, false);
-    }
+    await removeCourseFromDistributions(course);
```
Comment (mpark63) [inline routes/course.js:292]: remove course from all distributions it is a part of

File: routes/distribution.js [hunk 16]
```diff
@@ -6,67 +6,54 @@ import {
...
 import Distributions from '../model/Distribution.js';
 import Plans from '../model/Plan.js';
+import Majors from '../model/Major.js';
 import express from 'express';
+import { initDistributions } from './distributionMethods.js';
 
 const router = express.Router();
 
 //get distribution by id
 router.get('/api/distributions/:distribution_id', auth, async (req, res) => {
```
Comment (mpark63) [inline routes/distribution.js:18]: used when click on a distribution bar

File: routes/distribution.js [hunk 17]
```diff
@@ -6,67 +6,51 @@ import {
...
 });
 
 //get all distributions in a plan
-router.get('/api/distributionsByPlan/:plan_id', auth, async (req, res) => {
-  const plan_id = req.params.plan_id;
+router.get('/api/distributionsByPlan/', async (req, res) => {
+  const plan_id = req.query.plan_id;
+  const major_id = req.query.major_id;
+  if (!plan_id || !major_id) return missingHandler(res, { plan_id, major_id });
+  const reload = req.query.reload;
```
Comment (mpark63) [inline routes/distribution.js:37]: returns all existing distributions for specified plan / major 

if reload === 'true', re initializes distributions

File: routes/distribution.js [hunk 18]
```diff
@@ -6,67 +6,51 @@ import {
...
-
-//create distribution and update its plan
-router.post('/api/distributions', auth, async (req, res) => {
-  const distribution = req.body;
-  if (!distribution || Object.keys(distribution).length == 0) {
-    return missingHandler(res, { distribution });
-  }
-  if (distribution.user_id !== req.user._id) {
-    return forbiddenHandler(res);
-  }
```
Comment (mpark63) [inline routes/distribution.js:58]: no need for this route as we create distributions when add plan / major

File: routes/distribution.js [hunk 19]
```diff
@@ -77,7 +61,7 @@ router.patch(
   async (req, res) => {
     const required = req.query.required;
     const id = req.query.id;
-    if (!required || !id) {
+    if (Number.parseInt(required) === NaN || !id) {
```
Comment (mpark63) [inline routes/distribution.js:64]: PATCH api/distributions/updateRequiredCredits not used atm. 

could be useful in the future though? for customizable distributiosn

File: routes/distribution.js [hunk 20]
```diff
@@ -115,37 +100,7 @@ router.patch('/api/distributions/updateName', auth, async (req, res) => {
     await distribution.save();
     returnData(distribution, res);
   } catch (err) {
-    errorHandler(res, 400, err);
-  }
-});
-
-//delete a distribution and its associated courses(if that is the only distirbution the course belongs to) and update it's plan
-//return the deleted courses
-router.delete('/api/distributions/:d_id', auth, async (req, res) => {
```
Comment (mpark63) [inline routes/distribution.js:124]: no need for route, since delete distributiosn with plan / major

File: routes/major.js [hunk 21]
```diff
@@ -13,6 +13,17 @@ router.get('/api/majors/all', async (req, res) => {
   }
 });
 
+router.get('/api/majors/:major_id', async (req, res) => {
```
Comment (mpark63) [inline routes/major.js:16]: not used, but made one as we're using the majors collection more.

File: routes/plan.js [hunk 22]
```diff
@@ -87,46 +91,52 @@ router.post('/api/plans', auth, async (req, res) => {
...
-  if (plan.user_id !== req.user._id) {
+  if (planBody.user_id !== req.user._id) {
     return forbiddenHandler(res);
   }
   try {
-    const retrievedPlan = await Plans.create(plan);
+    const plan = await Plans.create(planBody);
     //update user
+    // add distributions for selected major(s)
+    await addPlanDistributions(plan);
```
Comment (mpark63) [inline routes/plan.js:112]: this is where we add empty distributions and finerequirements for a new plan

File: routes/plan.js [hunk 23]
```diff
@@ -193,31 +203,37 @@ router.delete('/api/plans/:plan_id', auth, async (req, res) => {
...
-    plan = await Plans.findByIdAndUpdate(id, updateBody, {
+    plan = await Plans.findByIdAndUpdate(plan_id, updateBody, {
       new: true,
       runValidators: true,
     }).exec();
-    const reviewers = await Reviews.find({ plan_id: id })
+    if (majors) {
+      await addPlanDistributions(plan);
+      await removePlanDistributions(plan);
+    }
```
Comment (mpark63) [inline routes/plan.js:233]: adds / removes distributions as necessary.

File: routes/search.js [hunk 24]
```diff
@@ -29,12 +30,45 @@ router.get('/api/search/skip/:num', async (req, res) => {
...
+      else if (number.includes('EN.550'))
+        number = number.replace('EN.550', 'EN.553');
+      courses = await SISCV.find({ number }).exec();
+      if (courses.length == 0) {
+        return returnData(-1, res);
+      }
+    }
+    returnData(courses[0], res);
+  } catch (err) {
+    errorHandler(res, 400, err);
```
Comment (mpark63) [inline routes/search.js:71]: i think this is in dev as well. some search changes got mixed in

File: routes/year.js [hunk 25]
```diff
@@ -144,12 +143,7 @@ router.delete('/api/years/:year_id', auth, async (req, res) => {
     year.courses.forEach(async (c_id) => {
       const course = await Courses.findByIdAndDelete(c_id).exec();
       course.distribution_ids.forEach(async (id) => {
-        const distribution = await Distributions.findByIdAndUpdate(
-          id,
-          { $pull: { courses: c_id } },
-          { new: true, runValidators: true },
-        ).exec();
-        await distributionCreditUpdate(distribution, course, false);
+        await removeCourseFromDistributions(course);
```
Comment (mpark63) [inline routes/year.js:146]: remove each course in year from distributions

File: tests/routes/course.spec.js
Comment (mpark63) [inline tests/routes/course.spec.js:1]: testing is definitely WIP ;-; idk why it keeps breaking

**Review by QinyuanWu — APPROVED** [review]

**Pull request closed without merging**